src/MongoDB.Bson/IO/JsonReader.cs

```diff
@@ -956,19 +965,117 @@ private BsonValue ParseBinDataExtendedJson()
+            {
+                if (nextToken.Type != JsonTokenType.String && nextToken.Type != JsonTokenType.UnquotedString)
+                {
+                    var message = string.Format("JSON reader expected a string but found '{0}'.", nextToken.Lexeme);
+                    throw new FormatException(message);
+                }
+                var name = nextToken.StringValue;
+
+                nextToken = PopToken();
+                if (nextToken.Type != JsonTokenType.Colon)
+                {
+                    var message = string.Format("JSON reader expected ':' but found '{0}'.", nextToken.Lexeme);
+                    throw new FormatException(message);
+                }
+
+                nextToken = PopToken();
+                if (nextToken.Type != JsonTokenType.String)
+                {
+                    var message = string.Format("JSON reader expected a string but found '{0}'.", nextToken.Lexeme);
+                    throw new FormatException(message);
+                }
+                var value = nextToken.StringValue;
+
+                switch (name)
+                {
+                    case "base64": base64String = value; break;
+                    case "subType": subTypeString = value; break;
+                    default:
+                        var message = string.Format("JSON reader expected 'base64' or 'subType', but found '{0}'.", name);
+                        throw new FormatException(message);
+                }
+
+                nextToken = PopToken();
+                if (nextToken.Type != JsonTokenType.Comma && nextToken.Type != JsonTokenType.EndObject)
+                {
+                    var message = string.Format("JSON reader expected ',' or '}}' but found '{0}'.", nextToken.Lexeme);
+                    throw new FormatException(message);
+                }
+
+                if (nextToken.Type == JsonTokenType.Comma)
+                {
+                    nextToken = PopToken();
+                }
+            }
+
+            if (base64String == null)
+            {
+                var message = "JSON reader expected $binary to contain a 'base64' element.";
                 throw new FormatException(message);
             }
-            var bytes = Convert.FromBase64String(bytesToken.StringValue);
+            if (subTypeString == null)
+            {
+                var message = "JSON reader expected $binary to contain a 'subType' element.";
+                throw new FormatException(message);
+            }
+            if (!IsValidBinaryDataSubTypeString(subTypeString))
+            {
+                var message = string.Format("JSON reader expected subType to be a one or two digit hex string, but found '{0}'.", subTypeString);
+                throw new FormatException(message);
+            }
+
+            bytes = Convert.FromBase64String(base64String);
+            subType = (BsonBinarySubType)Hex.ParseInt32(subTypeString);
```

Comment:
Which behavior is expected in case we have `subTypeString` value which cannot be converted to `BsonBinarySubType`?

Reply:
If `subTypeString` is not a valid 1-2 digit hex string a `FormatException` is thrown.

Any integer can be converted to a `BsonBinarySubType` even if the enum doesn't define a keyword for a particular value. That's legal.

Reply:
>even if the enum doesn't define a keyword for a particular value.
I asked about this part.
Ok, if it's legal